### ISSUE TITLE
SOAP Error

### ISSUE THREAD

**Issue by cdCorey** [skyverge/woocommerce-sage-erp-connector#12]
I'm getting this error:

> Export Failure - Order 2468 (ERP Sales Order No: N/A), Message : SOAP-ERROR: Parsing WSDL: Couldn't load from 'https://_mydomain.com_/eBusinessWebServices/MasBasicWsdl.svc?wsdl' : failed to load external entity "https://_mydomain.com_/eBusinessWebServices/MasBasicWsdl.svc?wsdl" (N/A)
> Order failed export to Sage ERP.

I saw issue #2, but I've confirmed my endpoint URL is accessible from the machine I'm testing on.

I found [this StackOverflow thread](http://stackoverflow.com/questions/12875409/soap-php-fault-parsing-wsdl-failed-to-load-external-entity) reporting this error can be caused by a change to SSL handling in PHP 5.6+, so I rolled back to PHP 5.4, but that didn't help. I also confirmed `allow_url_fopen` is on.

I found a [blog post](http://thinkglobal.co/resources/solved-soap-error-parsing-wsdl-failed-to-load-external-entity/) that reports encountering this error with a Magento API due to Magento referencing its own WSDL with an internal domain. When I access the WSDL URL in my browser, I can see this in the resulting XML dump:

```
<wsdl:types>
   <xsd:schema targetNamespace="http://mas90.sage.com/ws/Imports">
      <xsd:import schemaLocation="https://_workstation.workgroup.local_/eBusinessWebServices/MasBasicWsdl.svc?xsd=xsd0" namespace="http://mas90.sage.com/ws"/>
      <xsd:import schemaLocation="https://_workstation.workgroup.local_/eBusinessWebServices/MasBasicWsdl.svc?xsd=xsd1" namespace="http://schemas.microsoft.com/2003/10/Serialization/"/>
      <xsd:import schemaLocation="https://_workstation.workgroup.local_/eBusinessWebServices/MasBasicWsdl.svc?xsd=xsd2" namespace="http://schemas.microsoft.com/2003/10/Serialization/Arrays"/>
   </xsd:schema>
</wsdl:types>
```

I've rarely worked with SOAP, but I'm guessing this is the same issue that post is talking about. So I followed their recommendation of adding the internal workgroup domain to my external test machine's hosts file, but I'm still getting the same error. However, attempting to access said URLs via HTTPS at their internal domain in my browser generates an SSL error (as one would expect), which I imagine would also interfere with the plugin's connection.

I'd appreciate any insight you might be able to offer on this. Do we need to change some configuration somewhere (Sage? IIS?) to get the WSDL to point to the external domain? Or am I on the wrong track here?

**Comment by maxrice** [comment]
yeah, this is an issue with the IIS configuration for the Sage service. It's been a while since I've worked with it, but when viewing the WSDL in your browser you want to make sure that all the URLs reference the publicly-accessible endpoint, _not_ the internal server name. I can't remember the exact configuration settings but it's definitely within the IIS config somewhere, maybe the bindings?

**Comment by cdCorey** [comment]
Thanks, @maxrice. We got the domain name issue straightened out, but I'm still receiving the same error. At the moment I'm pretty much out of ideas. If you can offer any other suggestions of what might cause this error, or places I should be looking at to resolve this, I'd certainly appreciate it.

**Comment by maxrice** [comment]
You may want to find an IIS expert on upwork.com or similar to help troubleshoot further in that case. Can you access the WSDL from the internal network? If so, then you'll want to figure out if the internal firewall is correctly allowing external access to the IIS site.

**Comment by cdCorey** [comment]
The weird thing is I can access the /eBusinessWebServices/MasBasicWsdl.svc?wsdl URL from my browser, and it looks like a valid XML file to me.

**Comment by maxrice** [comment]
Hm, that is pretty weird indeed -- the last thing you could try is a SOAP tool locally on your computer to try and access the WSDL. If that works, then it would point to PHP as a potential issue, though it could be that the server where your site is hosted can't access the IIS site. So, you'll want to try and narrow down where you can access that WSDL and where you can't, to try and find the actual source of the issue.

**Comment by cdCorey** [comment]
@maxrice, turns out it was most likely a configuration issue with PHP, as you suggested. I've worked around that and I'm able to connect to Sage, but I'm now getting errors about the lack of a "shipping phone" and a "tax schedule." I am not aware of either of these fields existing in WooCommerce. We do have a sales tax rate entered into WooCommerce for the state our client operates out of, but I'm assuming if that's what it was looking for the plugin would have pulled that in automatically. Am I missing something in terms of configuration, or is there expected to be custom coding required to get this plugin functioning?

Thanks again for all your advice thus far. I understand this is officially an unsupported plugin, so I really appreciate your taking the time to offer assistance.

**Comment by maxrice** [comment]
Yeah there may be some custom coding required here since every Sage installation can be slightly different in the fields that it requires when creating customers/sales orders. If these are sales order fields it's complaining about, you can use [this filter](https://github.com/skyverge/woocommerce-sage-erp-connector/blob/master/classes/class-wc-sage-erp-connector-exporter.php#L297) to add the additional fields there. Shipping phone is straightforward and tax schedule is usually just a number IIRC. Try creating a sales order via the Sage UI and it should indicate what it considers valid entries for those fields.

**Comment by cdCorey** [comment]
Thanks, that makes sense.

--
Corey Caswick
Chatman Design
chatmandesign.com <http://www.chatmandesign.com>
